### PR TITLE
chore: skip klona for eval updates (#27048)

### LINKED ISSUE #27048: Phase 2 of large dataset performance optimisation: skip klona operations for evaluation updates
Has been observed that klona(deepClone) operations are performed on the entire dataTree for every evaluation. This computation increases with larger datasets, skipping this computation improves evalThread computation. It has been observed that eval scripting falls by almost 40% for collections for operations on datasets of size about 40000 records.

### PR DESCRIPTION
**Pull request by vsvamsi1**
## Description
Klona operations are being performed on the dataTree for every evaluation cycle. This computation is heavy and scales up for large dataset. Over here we optimise code to limit klona deepClones for every evaluation cycles and noticed reduction of eval scripting time of upto 40% for apps hosting about 40000 records. Seen a 20-25% reduction in eval scripting in MISC test cases, these test cases host about 20000 records.
#### PR fixes following issue(s)
Fixes #27048

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)
#### How Has This Been Tested?

- [x] Manual
- [ ] JUnit
- [ ] Jest
- [x] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 1]
```diff
@@ -928,18 +930,22 @@ export default class DataTreeEvaluator {
...
+      shouldNotCloneUnevalTree: false,
     },
     oldConfigTree: ConfigTree,
   ): {
     evaluatedTree: DataTree;
     evalMetaUpdates: EvalMetaUpdates;
     staleMetaIds: string[];
   } {
-    const tree = klona(oldUnevalTree);
+    const tree = options.shouldNotCloneUnevalTree
```
Comment (arunvjn) [inline app/client/src/workers/common/DataTreeEvaluator/index.ts:946]: Instead of introducing a new option, can you clone explicitly in the caller and pass the tree ? Optionally you could use the `isFirstTree` option as well.

Comment (vsvamsi1) [reply app/client/src/workers/common/DataTreeEvaluator/index.ts]: Sure

**Review by Rishabh-Rathod — DISMISSED** [review]
LGTM. How are we testing this change?

**Comment by vsvamsi1** [comment]
> LGTM. How are we testing this change?

I have requested @laveena-en to perform a deep regression on this, are there any edge cases that we need to specifically test?

**Comment by laveena-en** [comment]
Verified. Changes look good to me.

**Review by ohansFavour — APPROVED** [review]
Looks good

**Comment by vsvamsi1** [comment]
Seen atleast 20-30% drop in eval scripting results for MISC test cases which host about 20,000 records. For 40,000 records seen a drop in 40% in eval scripting
Screenshot 2023-09-15 at 3 29 45 PM
